src/kernel.ts

```diff
@@ -1,1069 +1,803 @@
-        /**
-         * Handle a failed AJAX request by logging the error message, and
-         * then optionally calling a provided callback.
-         *
-         * @function _on_error
-         * @param {function} error - callback
-         */
-        return (xhr, status, err) => {
-            utils.log_ajax_error(xhr, status, err);
-            if (error) {
-                error(xhr, status, err);
-            }
-        };
-    }
+/**
+ * Object providing a Future interface for message callbacks.
+ *
+ * Only one callback can be registered per type.
+ * If `autoDispose` is set, the future will self-dispose after `isDone` is
+ * set and the registered `onDone` handler is called.
+ *
+ * The Future is considered done when a `reply` message and a
+ * an `idle` iopub status message have been received.
+ */
+export
+interface IKernelFuture extends IDisposable {
+  /**
+   * The autoDispose behavior of the future.
+   *
+   * If True, it will self-dispose() after onDone() is called.
+   */
+  autoDispose: boolean;
+
+  /**
+   * Set when the message is done.  
+   */
+  isDone: boolean;
+
+  /**
+   * Register a reply handler. Returns `this`.
+   */
+  onReply(cb: (msg: IKernelMsg) => void): IKernelFuture;
+
+  /**
+   * Register an output handler. Returns `this`.
+   */
+  onOutput(cb: (msg: IKernelMsg) => void): IKernelFuture;
+
+  /**
+   * Register a done handler. Returns `this`.
+   */
+  onDone(cb: (msg: IKernelMsg) => void): IKernelFuture;
+
+  /**
+   * Register an input handler. Returns `this`.
+   */
+  onInput(cb: (msg: IKernelMsg) => void): IKernelFuture;
+}
 
-    _kernel_created(data): void {
-        /**
-         * Perform necessary tasks once the kernel has been started,
-         * including actually connecting to the kernel.
-         *
-         * @function _kernel_created
-         * @param {Object} data - information about the kernel including id
-         */
-        this.id = data.id;
-        this.kernel_url = utils.url_join_encode(this.kernel_service_url, this.id);
-        this.start_channels();
-    }
 
-    _kernel_connected(): void {
-        /**
-         * Perform necessary tasks once the connection to the kernel has
-         * been established. This includes requesting information about
-         * the kernel.
-         *
-         * @function _kernel_connected
-         */
-        events.trigger('kernel_connected.Kernel', {kernel: this});
-        // get kernel info so we know what state the kernel is in
-        var that = this;
-        this.kernel_info((reply?) => {
-            this.info_reply = reply.content;
-            events.trigger('kernel_ready.Kernel', {kernel: that});
+/**
+ * A Kernel class to communicate with the Python kernel. This
+ * should generally not be constructed directly, but be created
+ * by the `Session` object. Once created, this object should be
+ * used to communicate with the kernel.
+ */
+export
+class Kernel {
+
+  static statusChange = new Signal<Kernel, string>();
```

Comment:
I like to name signals in past tense, as in "something happened". So, this would be `statusChanged`.